Switch `#[track_caller]` back to a no-op unless feature gate is enabled

bryangarza opened the pull request:
This patch fixes a regression, in which `#[track_caller]`, which was previously a no-op, was changed to actually turn on the behavior. This should instead only happen behind the `closure_track_caller` feature gate.

Also, add a warning for the user to understand how their code will compile depending on the feature gate being turned on or not.

Fixes #104588

bryangarza committed:
Switch `#[track_caller]` back to a no-op unless feature gate is enabled

This patch fixes a regression, in which `#[track_caller]`, which was
previously a no-op, was changed to actually turn on the behavior. This
should instead only happen behind the `closure_track_caller` feature
gate.

Also, add a warning for the user to understand how their code will
compile depending on the feature gate being turned on or not.

Fixes #104588

bryangarza commented:
r? @compiler-errors

bryangarza commented on src/test/ui/async-await/track-caller/issue-104588-no-op-track-caller.stderr:
```diff
@@ -0,0 +1,12 @@
+LL |   #[track_caller]
+   |  ________________^
+LL | | async fn foo() {}
+   | |_
```
Not a huge fan of how the warning looks, for example the ascii arrow points to nothing (there's no text on this line). If someone has a suggestion on how to make this look cleaner overall, please let me know :)

estebank replied:
You also managed to find a bug in the output here! it should have been `|^` instead of `|_` 🤔

bryangarza replied:
oh! interesting haha. yeah i was like "this look weird..."

bryangarza replied:
I'll file a bug for this...

bryangarza replied:
Filed #105430

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3527319034/jobs/5916240477) [(plain)](https://github.com/rust-lang/rust/commit/e589c062de9a415c8f64d0ddb44e9e020ceda300/checks/9652266384/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
doc tests for: /checkout/src/doc/rustc/src/tests/index.md
doc tests for: /checkout/src/doc/rustc/src/what-is-rustc.md
 finished in 1.769 seconds
Generating lint docs (x86_64-unknown-linux-gnu)
error: failed to test example in lint docs for `ungated_async_fn_track_caller` in /checkout/compiler/rustc_lint/src/builtin.rs:1344: lint docs should contain the line `### Example`
This error was generated by the lint-docs tool.
This tool extracts documentation for lints from the source code and places
This tool extracts documentation for lints from the source code and places
them in the rustc book. See the declare_lint! documentation
https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lint_defs/macro.declare_lint.html
for an example of the format of documentation this tool expects.


To re-run these tests, run: ./x.py test --keep-stage=0 src/tools/lint-docs
The --keep-stage flag should be used if you have already built the compiler

```

</details>

compiler-errors commented on compiler/rustc_lint/src/builtin.rs:
```diff
@@ -1337,6 +1341,45 @@ impl<'tcx> LateLintPass<'tcx> for UnstableFeatures {
+            if let Some(attr) = maybe_track_caller {
+                cx.struct_span_lint(
+                    UNGATED_ASYNC_FN_TRACK_CALLER,
+                    span.with_hi(attr.span.hi()),
```
Why not just use `attr.span` here? What's the span you're trying to achieve?

bryangarza replied:
initially, I was just using `span` but that was printing the function but not the annotation. So this was how I got the annotation itself to also print out

compiler-errors replied:
This is a span that goes from the *end* of the attr to the *beginning* of the function item, and I don't think that's what you want even though it renders the correct lines partially.

You should just use `attr.span`, and then perhaps on line 1376 you can add a label using `lint.span_label` to also point out the function body.

compiler-errors commented on compiler/rustc_ast_lowering/src/expr.rs:
```diff
@@ -655,33 +655,40 @@ impl<'hir> LoweringContext<'_, 'hir> {
+        let hir_id = if self.tcx.features().closure_track_caller {
+            let parent_has_track_caller = self
+                .attrs
+                .values()
```
This is from your previous PR, but why is this correct? Shouldn't we be checking the attributes of only the enclosing item, not all enclosing items recursively?

bryangarza replied:
if i remember correctly, what i was trying to do was go one level up in the span hierarchy, because the attribute was attached to the parent span, not the span that is passed into this function. but maybe there is a better way to achieve that to avoid the nested find

bryangarza replied:
I tried printing out `self.attrs` here (`self` being the `LoweringContext`) and I get this (for the `panic-track-caller.rs` test)
```
INFO rustc_ast_lowering::expr self.attrs ⚠️  are SortedMap { data: [] }
INFO rustc_ast_lowering::expr self.attrs ⚠️  are SortedMap { data: [] }
INFO rustc_ast_lowering::expr self.attrs ⚠️  are SortedMap { data: [(0, [Attribute { kind: Normal(NormalAttr { item: AttrItem { path: Path { span: /local/home/garzbrya/code/rs/rust2/src/test/ui/async-await/track-caller/panic-track-caller.rs:50:3: 50:15 (#0), segments: [PathSegment { ident: track_caller#0, id: NodeId(179), args: None }], tokens: None }, args: Empty, tokens: None }, tokens: None }), id: AttrId(3), style: Outer, span: /local/home/garzbrya/code/rs/rust2/src/test/ui/async-await/track-caller/panic-track-caller.rs:50:1: 50:16 (#0) }])] }
INFO rustc_ast_lowering::expr self.attrs ⚠️  are SortedMap { data: [] }
```
Which makes me think that we have a separate `LoweringContext` per function in the test? Is that understanding correct? Or is `LoweringContext` encompassing all of the AST?

---

Ideally, there would be some span or ID that represents the entire async fn including the attributes, but I'm not able to find it in the code. At this point in the lowering it seems we are multiple layers removed from the attributes and I haven't found where exactly they are set yet. (Other than being available in `self.attrs`).

bryangarza replied:
I found where the `parse_fn` code is so I'm going to trace the logic from there and see if that helps me understand

bryangarza replied:
No longer needed (hirId check addressed in master branch already)

compiler-errors commented on compiler/rustc_ast_lowering/src/expr.rs:
```diff
@@ -655,33 +655,40 @@ impl<'hir> LoweringContext<'_, 'hir> {
+            let parent_has_track_caller = self
+                .attrs
+                .values()
+                .find(|attrs| {
+                    attrs.into_iter().find(|attr| attr.has_name(sym::track_caller)).is_some()
+                })
+                .is_some();
```
nit: `find + is_some = any`

```suggestion
                .any(|attrs| {
                    attrs.iter().any(|attr| attr.has_name(sym::track_caller))
                });
```

bryangarza replied:
thanks :) much cleaner!

compiler-errors commented on compiler/rustc_ast_lowering/src/expr.rs:
```diff
@@ -655,33 +655,40 @@ impl<'hir> LoweringContext<'_, 'hir> {
+                })
+                .is_some();
+            if parent_has_track_caller {
+                let generator_hir_id = self.lower_node_id(closure_node_id);
```
We call `lower_node_id(closure_node_id)` three times here -- is it possible to factor that out?

bryangarza replied:
will add in next commit

compiler-errors commented on compiler/rustc_error_messages/locales/en-US/lint.ftl:
```diff
@@ -350,6 +350,8 @@ lint_builtin_mutable_transmutes =
 
 lint_builtin_unstable_features = unstable feature
 
+lint_ungated_async_fn_track_caller = `#[track_caller]` on async functions is a no-op, unless the `closure_track_caller` feature is enabled
```
I wonder if this is better split into a primary message "... is a no-op" and a label that we only show if we're on nightly "... enable the `closure_track_caller` feature to ...".

What do you think?

bryangarza replied:
yeah that might be better, maybe something like

```
lint_ungated_async_fn_track_caller = `#[track_caller]` on async functions is a no-op
    .suggestion = enable the `closure_track_caller` feature
```
🤔
i'll look into how to only include the second part on nightly

bryangarza replied:
will add in next commit

compiler-errors commented on compiler/rustc_lint/src/builtin.rs:
```diff
@@ -1337,6 +1341,45 @@ impl<'tcx> LateLintPass<'tcx> for UnstableFeatures {
+        span: Span,
+        hir_id: HirId,
+    ) {
+        if let HirFnKind::ItemFn(_, _, _) = fn_kind && fn_kind.asyncness() == IsAsync::Async && !cx.tcx.features().closure_track_caller {
```
Is this `::ItemFn` check necessary? We could presumably also have the same situation on `async` closures and methods, too?

bryangarza replied:
you might be right... i need to test it out to make sure

bryangarza replied:
I removed the first condition

compiler-errors commented on compiler/rustc_lint/src/builtin.rs:
```diff
@@ -1337,6 +1341,45 @@ impl<'tcx> LateLintPass<'tcx> for UnstableFeatures {
+                    span.with_hi(attr.span.hi()),
+                    fluent::lint_ungated_async_fn_track_caller,
+                    |lint| lint,
+                    );
```
nit: indentation

```suggestion
                );
```

compiler-errors commented on src/test/ui/async-await/track-caller/issue-104588-no-op-panic-track-caller.rs:
```diff
@@ -0,0 +1,78 @@
+// run-pass
```
Is it possible to use `// revisions` to test this with and without the feature gate? Or is there any code that's exercising this feature with `closure_track_caller` enabled?

Otherwise I'm not sure I understand what this code is supposed to be testing :thinking:

bryangarza replied:
it's a copy paste of that other test with a couple lines changed

and the other tiny test i added just tests compilation so this one tests that the runtime behavior is as we expect.

but maybe this test isn't needed?

compiler-errors replied:
> it's a copy paste of that other test with a couple lines changed

Depending on what lines you changed, you may be able to use revisions instead: https://rustc-dev-guide.rust-lang.org/tests/compiletest.html#revisions

bryangarza replied:
I got it to work with revisions now instead of the separate file

compiler-errors reviewed (changes requested):
Thanks for the PR. I had some questions, though I may be missing some context, so feel free to correct me if I assumed something incorrectly.

bryangarza committed:
Update code based on PR comments

This patch does the following:
- Refactor some repeated lines into a single one
- Split the `ungated_async_fn_caller` lint into multiple lines, and make
  one of those lines only print out on nightly
- Use test revisions instead of copying an existing test

bryangarza committed:
update wording of lint

bryangarza commented:
I addressed most of the comments and resolved the thread, but there's a couple more I have to follow up with in another commit :)

bryangarza commented:
TODO for myself: add a doc comment to the lint definition

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3537551197/jobs/5937615348) [(plain)](https://github.com/rust-lang/rust/commit/c0ff47bceee9e3d8e8994bf9ce42b40949df452b/checks/9680111884/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
doc tests for: /checkout/src/doc/rustc/src/tests/index.md
doc tests for: /checkout/src/doc/rustc/src/what-is-rustc.md
 finished in 1.936 seconds
Generating lint docs (x86_64-unknown-linux-gnu)
error: failed to test example in lint docs for `ungated_async_fn_track_caller` in /checkout/compiler/rustc_lint/src/builtin.rs:1344: lint docs should contain the line `### Example`
This error was generated by the lint-docs tool.
This tool extracts documentation for lints from the source code and places
This tool extracts documentation for lints from the source code and places
them in the rustc book. See the declare_lint! documentation
https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lint_defs/macro.declare_lint.html


To re-run these tests, run: ./x.py test --keep-stage=0 src/tools/lint-docs
Build completed unsuccessfully in 0:34:50
The --keep-stage flag should be used if you have already built the compiler

```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #104863) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bryangarza commented:
My latest commit is near-identical to #105180 so I think I will need to just remove that one and rebase (there might be a couple lines to keep from the commit, but not related to the hirId)

bryangarza committed:
Update track_caller logic/lint after rebase

bryangarza committed:
Add lint doc comment

bryangarza commented:
@rustbot ready

compiler-errors commented on compiler/rustc_lint/src/builtin.rs:
```diff
@@ -1337,6 +1341,78 @@ impl<'tcx> LateLintPass<'tcx> for UnstableFeatures {
+                            "this function will not propagate the caller location",
+                        );
+                        if cx.tcx.sess.is_nightly_build() {
+                            lint.span_suggestion(
```
Hm, this suggestion is telling us to replace the span of the attribute with the replacement string `closure_track_caller`, i.e.:

```diff
- #[track_caller]
+ closure_track_caller
  async fn() {}
```

I'm not sure that's what we want to do here. You can pass the lint to [`add_feature_diagnostics`]() to annotate this with the standard missing nightly feature notes, and you don't need to gate this behind a nightly build check.

bryangarza replied:
thanks, I didn't know about that function!

bryangarza replied:
Updated in next commit

bryangarza commented:
Thanks for the comments! I’ll update this PR tomorrow

bors commented:
:umbrella: The latest upstream changes (presumably #105667) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bryangarza commented on compiler/rustc_ast_lowering/src/expr.rs:
```diff
@@ -678,30 +678,32 @@ impl<'hir> LoweringContext<'_, 'hir> {
-                        item: AttrItem {
-                            path: Path::from_ident(Ident::new(sym::track_caller, span)),
-                            args: AttrArgs::Empty,
+        if self.tcx.features().closure_track_caller {
+            let track_caller =
+                outer_hir_id.and_then(|id| self.attrs.get(&id.local_id)).map_or(false, |attrs| {
+                    attrs.into_iter().any(|attr| attr.has_name(sym::track_caller))
+                });
+            if track_caller {
```
good idea

bryangarza replied:
Updated in next commit

bryangarza committed:
Improve code based on feedback.

This patch improves the readability of some of the code by using
if-let-chains. Also, make use of the `add_feature_diagnostics` function.

bryangarza commented on compiler/rustc_lint/src/builtin.rs:
```diff
@@ -1337,6 +1341,78 @@ impl<'tcx> LateLintPass<'tcx> for UnstableFeatures {
+            // Now, check if the function has the `#[track_caller]` attribute
+            let attrs = cx.tcx.hir().attrs(hir_id);
+            let maybe_track_caller = attrs.iter().find(|attr| attr.has_name(sym::track_caller));
+            if let Some(attr) = maybe_track_caller {
```
Updated in next commit

bryangarza commented on compiler/rustc_lint/src/builtin.rs:
```diff
@@ -1337,6 +1341,78 @@ impl<'tcx> LateLintPass<'tcx> for UnstableFeatures {
+                    |lint| {
+                        lint.span_label(
+                            span,
+                            "this function will not propagate the caller location",
```
Updated in next commit

bryangarza committed:
Update track_caller tests; run fmt

bryangarza commented:
@rustbot ready

compiler-errors commented:
@bors r+ rollup

bors commented:
:pushpin: Commit ccbba0a60e3b094aeb48991cac9b6e342eb3e229 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors ccbba0a60e3b094aeb48991cac9b6e342eb3e229 -->
<!-- homu: {"type":"Approved","sha":"ccbba0a60e3b094aeb48991cac9b6e342eb3e229","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

compiler-errors commented:
Beta nominating this because this fixes a regression that's currently stable-to-beta where `#[track_caller]` on `async fn` turned from a no-op to a hard error. This instead emits a warning in that case for future compatibility reasons, but still allows the code to compile with the `track_caller` as a no-op.

apiraino commented:
Beta backport approved as per compiler team [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202022-12-22/near/317386930)

@rustbot label +beta-accepted

Mark-Simulacrum commented:
This PR does not backport cleanly -- @bryangarza would you be able to prepare a PR targeting the beta branch?

bryangarza commented:
Sounds good, I'll create the backport PR
